wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceValidator.java

```diff
@@ -166,6 +176,35 @@ private void validateLongAttribute(
     }
   }
 
+  /**
+   * Check that workspace has either plugins defined in attributes (Che 7) or installers defined in
+   * machines (Che 6).
+   *
+   * @throws ValidationException
+   */
+  private void validatePlugins(WorkspaceConfig config) throws ValidationException {
+    List<String> installers =
+        config
+            .getEnvironments()
+            .values()
+            .stream()
+            .map(
+                env ->
+                    env.getMachines()
+                        .values()
+                        .stream()
+                        .map(MachineConfig::getInstallers)
+                        .flatMap(List::stream))
+            .flatMap(Function.identity())
+            .collect(Collectors.toList());
```

Comment:
I think you don't need to collect all items and `findAny` will be enough here

Reply:
Yeah, it can be simpler a bit. 
```
config
            .getEnvironments()
            .values()
            .stream()
            .flatMap(env -> env.getMachines().values().stream()) //convert stream of envs to stream of machines
            .flatMap(machine -> machine.getInstallers())// convert to a stream of installers
            .blahblah() // find if there is at least 1 entry 
```
And also:
```
wsNext = attributes.containsKey(PLUGINS_ATTRIBUTE) || attributes.containsKey(EDITOR_ATTRIBUTE) 
```

Reply:
Thanks for the installers advice, I'll make that change. For the attributes, containsKey won't work because if e.g. a user toggles plugins on and off while creating a workspace it will add the attribute `"plugins"` with an empty key, and this can't be fixed during workspace creation.